File: monitoring-system/server/app/services/program_service.py
"""
Program service — CRUD for the ministry's daily schedule.
"""
from ..extensions import db
from ..models.program import Program


def list_programs(day: str = None, active_only: bool = True) -> tuple[dict, int]:
    query = Program.query
    if active_only:
        query = query.filter_by(is_active=True)
    if day:
        query = query.filter(Program.day.in_([day, "daily"]))
    programs = query.order_by(Program.start_time.asc()).all()
    return {"programs": [p.to_dict() for p in programs]}, 200


def create_program(data: dict) -> tuple[dict, int]:
    program = Program(**data)
    db.session.add(program)
    db.session.commit()
    return {"message": "Program created.", "program": program.to_dict()}, 201


def update_program(program_id: int, data: dict) -> tuple[dict, int]:
    program = Program.query.get_or_404(program_id)
    for key, value in data.items():
        setattr(program, key, value)
    db.session.commit()
    return {"message": "Program updated.", "program": program.to_dict()}, 200


def delete_program(program_id: int) -> tuple[dict, int]:
    program = Program.query.get_or_404(program_id)
    db.session.delete(program)
    db.session.commit()
    return {"message": "Program deleted."}, 200

File: monitoring-system/server/src/services/programService.js
import { prisma } from '../db.js';
import { HttpError } from '../utils/httpError.js';
import { programDict } from '../utils/serialize.js';
import { parseHHMM } from '../utils/dates.js';

function toRow(data) {
  return {
    title: data.title,
    description: data.description ?? '',
    day: data.day ?? 'daily',
    start_time: parseHHMM(data.start_time),
    end_time: parseHHMM(data.end_time),
    location: data.location ?? null,
    is_active: data.is_active ?? true,
  };
}

export async function listPrograms(day, activeOnly = true) {
  const where = {};
  if (activeOnly) where.is_active = true;
  if (day) where.day = { in: [day, 'daily'] };
  const programs = await prisma.programs.findMany({ where, orderBy: { start_time: 'asc' } });
  return { programs: programs.map(programDict) };
}

export async function createProgram(data) {
  const program = await prisma.programs.create({ data: toRow(data) });
  return { status: 201, body: { message: 'Program created.', program: programDict(program) } };
}

export async function updateProgram(id, data) {
  const existing = await prisma.programs.findUnique({ where: { id } });
  if (!existing) throw new HttpError(404, 'Resource not found.');
  const program = await prisma.programs.update({ where: { id }, data: toRow(data) });
  return { message: 'Program updated.', program: programDict(program) };
}

export async function deleteProgram(id) {
  const existing = await prisma.programs.findUnique({ where: { id } });
  if (!existing) throw new HttpError(404, 'Resource not found.');
  await prisma.programs.delete({ where: { id } });
  return { message: 'Program deleted.' };
}
